egui/src/widgets/drag_value.rs

```diff
@@ -157,130 +165,176 @@ impl<'a> Widget for DragValue<'a> {
-                    .font(TextStyle::Monospace),
-            );
-            if let Ok(parsed_value) = value_text.parse() {
-                let parsed_value = clamp_to_range(parsed_value, clamp_range);
-                set(&mut get_set_value, parsed_value);
-            }
-            if ui.input().key_pressed(Key::Enter) {
-                ui.memory().surrender_focus(kb_edit_id);
-                ui.memory().drag_value.edit_string = None;
-            } else {
-                ui.memory().drag_value.edit_string = Some(value_text);
-            }
-            response
-        } else {
-            let button = Button::new(
-                RichText::new(format!("{}{}{}", prefix, value_text, suffix)).monospace(),
-            )
-            .wrap(false)
-            .sense(Sense::click_and_drag())
-            .min_size(ui.spacing().interact_size); // TODO: find some more generic solution to `min_size`
-
-            let response = ui.add(button);
-            let mut response = response.on_hover_cursor(CursorIcon::ResizeHorizontal);
-
-            if ui.style().explanation_tooltips {
-                response = response .on_hover_text(format!(
-                    "{}{}{}\nDrag to edit or click to enter a value.\nPress 'Shift' while dragging for better control.",
-                    prefix,
-                    value as f32, // Show full precision value on-hover. TODO: figure out f64 vs f32
-                    suffix
-                ));
-            }
-
-            if response.clicked() {
-                ui.memory().request_focus(kb_edit_id);
-                ui.memory().drag_value.edit_string = None; // Filled in next frame
-            } else if response.dragged() {
-                ui.output().cursor_icon = CursorIcon::ResizeHorizontal;
-
-                let mdelta = response.drag_delta();
-                let delta_points = mdelta.x - mdelta.y; // Increase to the right and up
-
-                let speed = if is_slow_speed { speed / 10.0 } else { speed };
-
-                let delta_value = delta_points as f64 * speed;
-
-                if delta_value != 0.0 {
-                    let mut drag_state = std::mem::take(&mut ui.memory().drag_value);
-
-                    // Since we round the value being dragged, we need to store the full precision value in memory:
-                    let stored_value = (drag_state.last_dragged_id == Some(response.id))
-                        .then(|| drag_state.last_dragged_value)
-                        .flatten();
-                    let stored_value = stored_value.unwrap_or(value);
-                    let stored_value = stored_value + delta_value as f64;
-
-                    let aim_delta = aim_rad * speed;
-                    let rounded_new_value = emath::smart_aim::best_in_range_f64(
-                        stored_value - aim_delta,
-                        stored_value + aim_delta,
+        let mut response = ui
+            .horizontal(|ui| {
```

Comment:
This changes `DragValue` from being an atomic widget (which can e.g. be centered in a layout) to a composite widget, which is a bit of a shame. Do the step button really need to be part of the `DragValue` widget? Perhaps it would make the code simpler if there instead was a `DragStepValue` widget which consisted of a `DragValue` and two buttons? Seems more composable than just adding more features into the same widget.

Reply:
Good points, I extracted the buttons into a separate widget. Let me know what you think.

What's your opinion on using the unicode characters for the `-` and `+` buttons? They look nice when the font supports it but won't show correctly if it doesn't.

Reply:
I think using the fancy unicode characters makes sense for now - if it causes a problem down-the-road we can think about adding fallbacks.